core/src/main/java/org/elasticsearch/index/query/BoolQueryParser.java

```diff
@@ -71,14 +71,12 @@ public QueryBuilder fromXContent(QueryParseContext parseContext) throws IOExcept
                 switch (currentFieldName) {
                 case "must":
                     query = parseContext.parseInnerQueryBuilder();
-                    if (query != null) {
-                        mustClauses.add(query);
-                    }
+                    mustClauses.add(query);
                     break;
                 case "should":
                     query = parseContext.parseInnerQueryBuilder();
-                    if (query != null) {
-                        shouldClauses.add(query);
+                    shouldClauses.add(query);
+                    if (query != EmptyQueryBuilder.PROTOTYPE) {
                         if (parseContext.isFilter() && minimumShouldMatch == null) {
```

Comment:
makes sense, add a comment that explains just that? this stuff is tricky and we will forget about it